File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- Write your first then lower case function here
function first_then_lower_case(sequence, predicate)
  for _, value in ipairs(sequence) do
    if predicate(value) then return value:lower() end
  end
  return nil
 end

-- Write your powers generator here
function powers_generator(base, limit)
  return coroutine.create(function ()
    local value = 1
    while (value < limit) do
      coroutine.yield(value)
      value = value * base
    end
  end)
end

-- Write your say function here
function say(word)
  if word == nil then
    return ""
  end
  return function (next) return ((next == nil) and word or say(word .. " " .. next)) end
 end

-- Write your line count function here
function meaningful_line_count(filename)
  local line_number = 0
  local file = io.open(filename, 'r')
  if file == nil then
    error("No such file")
  end
  for line in file:lines() do
    local strippedline = string.gsub(line, " ", "")
    strippedline = string.gsub(strippedline, "\n", "")
    strippedline = string.gsub(strippedline, "\013", "")
    strippedline = string.gsub(strippedline, "\009", "")
    if ((strippedline ~= "") and (strippedline ~= nil) and (strippedline:sub(1,1) ~= '#')) then
      line_number = line_number + 1
    end
  end
  file:close()
  return line_number
 end
-- Write your Quaternion table here
Quaternion = {}
function Quaternion.new(a, b, c, d)
  local new_quat = {a=a, b=b, c=c, d=d}
  setmetatable(new_quat, Quaternion)
  Quaternion.__index = Quaternion
  return new_quat
end

function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end

function Quaternion:eq(other)
  return self.a == other.a and self.b == other.b and self.c == other.c and self.d == other.d
end

function Quaternion:add(other)
  return Quaternion.new(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)
end

function Quaternion:mul(other)
  local mul_a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
  local mul_b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
  local mul_c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
  local mul_d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
  return Quaternion.new(mul_a, mul_b, mul_c, mul_d)
end

function Quaternion:tostring(other)
  local function coeftostring(coef, comp)
    return ((coef >= 0) and "+" or "-") 
    .. ((math.abs(coef) ~= 1 or comp == "") and tostring(math.abs(coef)) or "") 
    .. comp
  end

  local components = {"", "i", "j", "k"}
  local stringified = ""
  for i, c in ipairs(self:coefficients()) do
    if c ~= 0 then
      stringified = stringified .. coeftostring(c, components[i])
    end
  end
  if (stringified == "") then
    return "0"
  end
  -- This is how you access individual characters in lua
  -- I dont make the rules
  if (string.sub(stringified, 1, 1) == "+") then
    stringified = string.sub(stringified, 2)
  end
  return stringified
end

Quaternion.__eq = Quaternion.eq
Quaternion.__add = Quaternion.add
Quaternion.__mul = Quaternion.mul
Quaternion.__tostring = Quaternion.tostring

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Optional, Generator, Union, overload


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(strings: list[str], predicate: Callable, /) -> Optional[str]     :
    for i in range(len(strings)):
        strings[i] = strings[i].lower()
        if predicate(strings[i]) == True:
            return strings[i]
    return None

# Write your powers generator here
def powers_generator(*, limit: int, base: int) -> Generator[int, None, None]:
    power = 0
    while True:
        result = base ** power
        if result > limit:
            break
        yield result
        power += 1

# Write your say function here
def say(message = None, /) -> Union[str, Callable]:
    if message == None:
        return ""
    else:
        # This effectively allows the function to instantiate a
        # new function that maintains the state of the message
        # being created
        class sayer:
            def __init__(self, message: Optional[str] = None):
                self.message = message
            def __call__(self, next: Optional[str] = None) -> Optional[Union[str, "sayer"]]:
                if type(next) is str and type(self.message) is str:
                    return sayer(self.message + " " + next)
                else:
                    return self.message
        return sayer(message)

# Write your line count function here
def meaningful_line_count(filename: str, /) -> int:
    line_count: int = 0
    with open(filename, mode='r', encoding="utf-8") as file:
        for line in file.readlines():
            stripped = line.strip()
            if stripped:
                if stripped[0] != '#':
                    line_count += 1
    return line_count


# Write your Quaternion class here

#Use a frozen dataclass DONE
#Overload the operators (+, *, and ==) DONE
#Require positional arguments DONE
#Make the conjugate and coefficients methods @propertys. NO IDEA
@dataclass(frozen=True)
class Quaternion:
    a: int
    b: int
    c: int
    d: int
    
    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)
    
    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, other: "Quaternion") -> "Quaternion":
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)

    def __mul__(self, other: "Quaternion") -> "Quaternion": 
        mul_a: int = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
        mul_b: int = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
        mul_c: int = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
        mul_d: int = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        return Quaternion(mul_a, mul_b, mul_c, mul_d)
    
    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Quaternion):
            return NotImplemented
        return (self.a == other.a) and (self.b == other.b) and (self.c == other.c) and (self.d == other.d)

    def __str__(self) -> str:
        # Forms the stringified number, including its sign and direction, if applicable, for each coefficient of the quaternion
        def coef_to_str(coef, comp, /):
            return ("+" if coef >= 0 else "-") + (str(abs(coef)) if abs(coef) != 1 or not comp else "") + comp

        components = ("", "i", "j", "k")
        coef_strings: list[str] = [coef_to_str(c, components[i]) for i, c in enumerate(self.coefficients) if c != 0]
        if not coef_strings:
            return "0"
        else:
            coef_string = "".join(coef_strings)
            return coef_string[1:] if coef_string[0] == '+' else coef_string